### ISSUE TITLE
Why does flashlight use CTC decoder in text2speech instead of a LAS-based (Listen, attend and spell) one?

### ISSUE THREAD

**Issue by gevorgyana** [flashlight/flashlight#575]
### Question
It happens that LAS does not assume conditional independence of the predicted tokens, so why not use that technique instead of CTC? LAS can theoretically substitute a language model by learning the structure of the language.

**Comment by tlikhomanenko** [comment]
We don't have implementation for LAS but we have seq2seq criterion so you can train it instead of CTC and get similar results as LAS. https://github.com/flashlight/flashlight/blob/master/flashlight/app/asr/criterion/TransformerCriterion.cpp, https://github.com/flashlight/flashlight/blob/master/flashlight/app/asr/common/Flags.cpp#L358. Also before we released some seq2seq models in https://github.com/flashlight/wav2letter/tree/master/recipes/sota/2019 and showed that actually with large scale experiments you have similar results for CTC and seq2seq after decoding.

**Comment by gevorgyana** [comment]
Excuse for me re-opening this, but I did not quite understand what seq2seq criterion is? I am sorry I have no clue. I know Alex Graves fought the problem of having to align inputs to outputs in his thesis paper and introduced CTC. By googling 'ctc alternative' I did not really find a paper talking about seq2seq criterion. It would be of much help if you could point me at a paper describing seq2seq criterion. Thank you!

**Comment by tlikhomanenko** [comment]
Please have a look at this reference. Sutskever, I., Vinyals, O., and Le, Q. V. Sequence to sequence learning with neural networks. InAdvances inneural information processing systems, pp. 3104–3112, 2014. It can e used also for speech as you want to map sequence of audio with one length to the sequence of tokens of another length.

**Comment by georgekald79** [comment]
> Please have a look at this reference. Sutskever, I., Vinyals, O., and Le, Q. V. Sequence to sequence learning with neural networks. InAdvances inneural information processing systems, pp. 3104–3112, 2014. It can e used also for speech as you want to map sequence of audio with one length to the sequence of tokens of another length.

**Comment by gevorgyana** [comment]
Okay, so general sequence to sequence as referred to in the paper is trained to minimize the distance between the correct output sequence and the predicted one, right?

I have also made some notes and hope that you correct me in case I misunderstood.

- "Sequence to sequence" does not use attention, it is simply 2 RNNs (one takes a sequence of input, one produces a sequence of output). There is a fixed-sized representation vector in between. It is somehow similar to RNN-T, but it does not involve a language model. It does not involve explicit attention either, only the implicit memory learned by the fixed-sized vector representation of the input. 

- In comparison to CTC, "sequence to sequence" still in theory is capable of modeling more context than CTC because at least it has a fixed-sized vector representation of its history through LSTMs, and CTC is supposed to give one character at a time, independent of its history. Am I correct?

**Comment by tlikhomanenko** [comment]
> Okay, so general sequence to sequence as referred to in the paper is trained to minimize the distance between the correct output sequence and the predicted one, right?

Sequence to sequence can work with assumption that input and output have different length. And you simply compute probability for each output token to be on this position and then compute log loss.

> I have also made some notes and hope that you correct me in case I misunderstood.
> 
>     * "Sequence to sequence" does not use attention, it is simply 2 RNNs (one takes a sequence of input, one produces a sequence of output). There is a fixed-sized representation vector in between. It is somehow similar to RNN-T, but it does not involve a language model. It does not involve explicit attention either, only the implicit memory learned by the fixed-sized vector representation of the input.
> 

Nope, this is not true. We have two implementations of seq2seq. One with RNN-decoder and another with Transformer-decoder. In both cases decoder attends to the encoder output vie attention mechanism and uses the previous context. So in some sense this is variation of classical RNN-T.

>     * In comparison to CTC, "sequence to sequence" still in theory is capable of modeling more context than CTC because at least it has a fixed-sized vector representation of its history through LSTMs, and CTC is supposed to give one character at a time, independent of its history. Am I correct?

Yep, CTC assumes independence of current position from previous positions. In seq2seq you use both all previously inferred context + the whole encoder output.

**Comment by gevorgyana** [comment]
Thank you very much, you gave me useful directions.
